src/cache.rs

```diff
@@ -4,144 +4,132 @@
+        } else {
+            db = DB::open_cf(&opts, path, list?)?;
         }
 
-        CACHE_DATA_PATH
-            .set(data_dir)
-            .expect("cache path can only be set once");
-
-        Persistence::with_backend(&backend(), |_| Ok(()))?;
+        let write_batch = WriteBatch::default();
 
-        Ok(())
+        Ok(Self { db, write_batch })
     }
 
-    /// Persist the cache to the internal backend.
-    pub(crate) fn persist(&self) -> Result<(), Error> {
-        let data_dir = CACHE_DATA_PATH.get().expect("cache path must be set");
-        let id_path = data_dir.join("cache.id");
-
-        let persistence_id = Persistence::persist(&backend(), &self.data)?;
-
-        let mut bytes = [0u8; 36];
-        let mut sink = Sink::new(&mut bytes);
-
-        persistence_id.encode(&mut sink);
-
-        fs::write(id_path, bytes)?;
-
-        Ok(())
-    }
-
-    /// Insert a note into the cache at the given block height. If note is
-    /// `None` only the block height gets updated.
+    // We store a column family named by hex representation of the psk.
+    // We store key as (psk, note) and value as (KeyData)
+    // The Canon of the tuple, we use it so we can have unique keys per psk and
+    // note
     pub(crate) fn insert(
         &mut self,
         psk: PublicSpendKey,
         height: u64,
-        note: Option<Note>,
+        note_data: (Option<Note>, Option<BlsScalar>),
     ) -> Result<(), Error> {
-        if self.data.get(&psk)?.is_none() {
-            self.data.insert(
-                psk,
-                KeyData {
-                    last_height: 0,
-                    notes: BTreeSet::new(),
-                },
-            )?;
+        let cf: &ColumnFamily;
+        let cf_name = format!("{:?}", psk);
+        let last_height_key = b"last_height";
+
+        if let Some(column_family) = self.db.cf_handle(&cf_name) {
+            cf = column_family;
+        } else {
+            // immediately create a cf by the hex of the psk_bytes
+            self.db.create_cf(&cf_name, &Options::default())?;
+
+            cf = self
+                .db
+                .cf_handle(&cf_name)
+                .expect("cannot create column family for db");
         }
 
-        let mut key_data = self.data.get_mut(&psk)?.expect("psk");
-
-        if height > key_data.last_height {
-            key_data.last_height = height;
+        if let Some(last_height) = self.db.get_cf(cf, last_height_key)? {
+            let last_height = u64::from_be_bytes(
+                last_height.try_into().expect("Invalid u64 in cache db"),
+            );
+
+            if height > last_height {
+                self.write_batch.put_cf(
+                    cf,
+                    last_height_key,
+                    height.to_be_bytes(),
+                );
+            }
+        } else {
+            self.write_batch
+                .put_cf(cf, last_height_key, height.to_be_bytes())
         }
```

Comment:
You can reuse the `last_height` method, it reduce the complexity and increase the reusability

```rust
        if height > self.last_height(psk)? {
            self.write_batch
                .put_cf(cf, last_height_key, height.to_be_bytes());
        }
```